### PR TITLE
patch to add throwing exception on 4XX/5XX error, instead of returning undef

### PR DESCRIPTION
**Pull request by jonathanstowe**
This is a rework of the https://github.com/cosimo/perl6-lwp-simple/pull/69 to add the option of throwing an exception on an error response.

Rather than doing it always this adds a switch to turn on the behaviour.

### CODE REVIEW

File: lib/LWP/Simple.pm [hunk 1]
```diff
@@ -87,17 +103,23 @@ method request_shell (RequestType $rt, Str $url, %headers = {}, Any $content?) {
...
-    # HTML header names can be mixed case. To find the desired names, transform
-    # them into lowercase.
-    my %resp_header_lowercase;
-    for $resp_headers.keys -> $header_name {
-        %resp_header_lowercase{$header_name.lc} = $resp_headers{$header_name};
-    }
-
     given $status {
 
+        when / <[4..5]> <[0..9]> <[0..9]> / {
```
Comment (lizmat) [inline lib/LWP/Simple.pm:108]: Wow, what a way to check `400 <= $_ <= 599`

File: lib/LWP/Simple.pm [hunk 2]
```diff
@@ -87,17 +103,23 @@ method request_shell (RequestType $rt, Str $url, %headers = {}, Any $content?) {
...
+                    headers => $resp_headers,
+                    content => self!decode-response( :$resp_headers, :$resp_content )
+                ).throw;
+            }
+            else {
+                return Nil;
+            }
+        }
+
         when / 30 <[12]> / {
```
Comment (lizmat) [inline lib/LWP/Simple.pm:121]: `301 | 302`

File: lib/LWP/Simple.pm [hunk 3]
```diff
@@ -113,37 +135,11 @@ method request_shell (RequestType $rt, Str $url, %headers = {}, Any $content?) {
         }
 
         when / 20 <[0..9]> / {
```
Comment (lizmat) [inline lib/LWP/Simple.pm:137]: `200 <= $_ <= 209`

**Review by lizmat — APPROVED** [review]

File: lib/LWP/Simple.pm (hunk 1 above)
Comment (jonathanstowe) [reply lib/LWP/Simple.pm:108]: Which I guess either are substantially quicker than a regex.

I think if we merge this, I'll fix it in another commit as this was a pre-existing condition :-)

File: lib/LWP/Simple.pm (hunk 3 above)
Comment (jonathanstowe) [reply lib/LWP/Simple.pm:137]: I'm actually tempted to make some subsets like `subset OKResponse where 200 <= * <= 209` just to make it a little neater.

**Comment by JJ** [comment]
Thanks a lot for this. I'll try to check it out next Monday.

File: lib/LWP/Simple.pm (hunk 3 above)
Comment (JJ) [reply lib/LWP/Simple.pm:137]: Waiting for that change...

Comment (jonathanstowe) [reply lib/LWP/Simple.pm:137]: It wants to be done as a separate piece of work as it was like it before this and the original PR was just following that.

**Comment by jonathanstowe** [comment]
Imma merge this now.

**Pull request merged**